test/upgrade/mzcompose.py

```diff
@@ -137,5 +181,192 @@ def test_upgrade_from_version(
+    )
+    schema_registry = SchemaRegistry(
+        depends_on=["kafka", "zookeeper", "test-certs"],
+        environment=[
+            "SCHEMA_REGISTRY_KAFKASTORE_TIMEOUT_MS=10000",
+            "SCHEMA_REGISTRY_HOST_NAME=schema-registry",
+            "SCHEMA_REGISTRY_LISTENERS=https://0.0.0.0:8081",
+            "SCHEMA_REGISTRY_KAFKASTORE_CONNECTION_URL=zookeeper:2181",
+            "SCHEMA_REGISTRY_KAFKASTORE_SECURITY_PROTOCOL=SSL",
+            "SCHEMA_REGISTRY_KAFKASTORE_SSL_KEYSTORE_LOCATION=/etc/schema-registry/secrets/schema-registry.keystore.jks",
+            "SCHEMA_REGISTRY_SSL_KEYSTORE_LOCATION=/etc/schema-registry/secrets/schema-registry.keystore.jks",
+            "SCHEMA_REGISTRY_KAFKASTORE_SSL_KEYSTORE_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_SSL_KEYSTORE_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_KAFKASTORE_SSL_KEY_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_SSL_KEY_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_KAFKASTORE_SSL_TRUSTSTORE_LOCATION=/etc/schema-registry/secrets/schema-registry.truststore.jks",
+            "SCHEMA_REGISTRY_SSL_TRUSTSTORE_LOCATION=/etc/schema-registry/secrets/schema-registry.truststore.jks",
+            "SCHEMA_REGISTRY_KAFKASTORE_SSL_TRUSTSTORE_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_SSL_TRUSTSTORE_PASSWORD=mzmzmz",
+            "SCHEMA_REGISTRY_SCHEMA_REGISTRY_INTER_INSTANCE_PROTOCOL=https",
+            "SCHEMA_REGISTRY_SSL_CLIENT_AUTH=true",
+        ],
+        volumes=[
+            "secrets:/etc/schema-registry/secrets",
+        ],
+        bootstrap_server_type="SSL",
     )
+    materialized = Materialized(
+        environment=[
+            "SSL_KEY_PASSWORD=mzmzmz",
+        ],
+        volumes_extra=["secrets:/share/secrets"],
+    )
+    testdrive = Testdrive(
+        entrypoint=[
+            "bash",
+            "-c",
+            "cp /share/secrets/ca.crt /usr/local/share/ca-certificates/ca.crt && "
+            "update-ca-certificates && "
+            "testdrive "
+            "--kafka-addr=kafka:9092 "
+            "--schema-registry-url=https://schema-registry:8081 "
+            "--materialized-url=postgres://materialize@materialized:6875 "
+            "--cert=/share/secrets/producer.p12 "
+            "--cert-password=mzmzmz "
+            "--ccsr-password=sekurity "
+            "--ccsr-username=materialize "
+            '"$$@"',
+        ],
+        volumes_extra=["secrets:/share/secrets"],
+        # Required to install root certs above
+        propagate_uid_gid=False,
+        validate_catalog=False,
+    )
+
+    with c.override(kafka, schema_registry, materialized, testdrive):
+        c.start_and_wait_for_tcp(services=["zookeeper", "kafka", "schema-registry"])
+        for version in tested_versions:
+            priors = [f"v{v}" for v in all_versions if v < version]
+            test_upgrade_from_version(c, f"v{version}", priors, filter=args.filter)
+
+
+def create_ssl_services() -> Tuple[Kafka, SchemaRegistry, Testdrive]:
```

Comment:
You may wish to remove the `create_` prefix from this method, as it is basically returning service definitions, it is not really creating the services themselves in the docker sense.